File: frontend/src/engine/capabilities/addHyperlink.ts
/**
 * addHyperlink – Insert a hyperlink in a cell.
 */

import { CapabilityMeta, AddHyperlinkParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "addHyperlink",
  description: "Insert a hyperlink in a cell",
  mutates: true,
  affectsFormatting: false,
  // Range.hyperlink was introduced in ExcelApi 1.7 (Excel 2019+).
  requiresApiSet: "ExcelApi 1.7",
};

async function handler(
  context: Excel.RequestContext,
  params: AddHyperlinkParams,
  options: ExecutionOptions
): Promise<StepResult> {
  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would add hyperlink to ${params.cell}` };
  }

  options.onProgress?.("Inserting hyperlink...");

  const range = resolveRange(context, params.cell);
  range.hyperlink = {
    address: params.url,
    textToDisplay: params.displayText ?? params.url,
  };
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Added hyperlink to ${params.cell}: ${params.url}`,
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.7) ────────────────────────────────────
// Range.hyperlink requires 1.7. The `=HYPERLINK(url, display)` worksheet function
// exists in every Excel version back to 97 and produces a clickable cell that
// navigates to the URL — visually and functionally identical to what the user
// asked for. Escape any embedded double-quotes in the url/display text so the
// formula stays well-formed even for odd URLs.
async function fallback(
  context: Excel.RequestContext,
  params: AddHyperlinkParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would add HYPERLINK() formula to ${params.cell}` };
  }

  options.onProgress?.("Legacy-Excel mode: writing HYPERLINK() formula instead of Range.hyperlink...");

  const escape = (s: string) => s.replace(/"/g, '""');
  const display = params.displayText ?? params.url;
  const formula = `=HYPERLINK("${escape(params.url)}","${escape(display)}")`;

  const range = resolveRange(context, params.cell);
  range.formulas = [[formula]];
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message:
      `Added HYPERLINK() formula to ${params.cell}: ${params.url} ` +
      `(legacy-Excel fallback — Range.hyperlink requires ExcelApi 1.7+).`,
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/add_hyperlink.py
"""addHyperlink — add a hyperlink to a cell."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("cell") or params.get("range")
    url = params.get("url") or params.get("address")
    display = params.get("displayText") or url
    if not address or not url:
        return {"status": "error", "message": "addHyperlink requires 'cell' and 'url'."}

    rng = resolve_range(ctx.workbook_handle, address)
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would add hyperlink to {rng.address}."}

    try:
        rng.sheet.api.Hyperlinks.Add(Anchor=rng.api, Address=url, TextToDisplay=display)
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Add hyperlink failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Added hyperlink to {rng.address}: {url}.",
        "outputs": {"cell": rng.address, "url": url},
    }


registry.register("addHyperlink", handler, mutates=True)
